### PR TITLE
feat(devops): Integration of PGBouncer for connection pooling

### PR DESCRIPTION
**Pull request by Vaish24navi**
## Description

Implemented `PgBouncer` to improve connection pooling and optimize database performance for the application. Ensured the following tasks are addressed:

 - Modified PostgreSQL settings to authenticate using MD5 for compatibility with `PgBouncer`.
-  Added `PgBouncer` to the existing Docker Compose setup.
-  Validated connections from the application to `PgBouncer` and then to PostgreSQL.

## Related Issue

- [X] Link to the related issue (if applicable) #48 

## Type of Change

- [X] New Feature

## Checklist

- [X] Code follows the project style guidelines
- [X] Tests have been added/updated
- [X] All tests pass

## Summary by CodeRabbit

- **New Features**
	- Enhanced logging for database connection attempts and success messages.
	- Introduced a new `pgbouncer` service for improved database connection management.

- **Configuration Changes**
	- Updated database configuration to retrieve values from environment variables.
	- Added new environment variables for the `app` and `pgbouncer` services in the Docker configuration.

- **Network Changes**
	- Switched network configuration from `app-network` to `backend` for both `app` and `db` services.
  
- **Parameter Update**
	- Changed default value of `DB_HOST` from `"localhost"` to `"127.0.0.1"`.

### CODE REVIEW

File: backend/app/utils/settings/config.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ class Settings(BaseSettings):
     DB_NAME: str = "datastore"
     DB_USER: str = "postgres"
     DB_PASSWORD: str = "postgres"
-    DB_HOST: str = "localhost"
+    DB_HOST: str = "127.0.0.1"
```
[the remark this answers is not in the document]
Comment (Vaish24navi) [reply backend/app/utils/settings/config.py:38]: but using localhost is causing DNS problems with PgBouncer

[the remark this answers is not in the document]
Comment (Vaish24navi) [reply backend/app/utils/settings/config.py]: resolved in the following commit
now all values are env-aware!

File: backend/app/utils/settings/config.py [hunk 2]
```diff
@@ -30,12 +33,12 @@ class Settings(BaseSettings):
         The internal table name.
     """
 
-    API_KEY: str = "OPEN_SESAME"
+    API_KEY: str = os.getenv("API_KEY")
```
[the remark this answers is not in the document]
Comment (Vaish24navi) [reply backend/app/utils/settings/config.py]: resolved in following commits

File: backend/app/utils/settings/config.py [hunk 3]
```diff
@@ -30,12 +33,12 @@ class Settings(BaseSettings):
...
 
-    DB_NAME: str = "datastore"
-    DB_USER: str = "postgres"
-    DB_PASSWORD: str = "postgres"
-    DB_HOST: str = "localhost"
+    DB_NAME: str = os.getenv("DB_NAME")
+    DB_USER: str = os.getenv("DB_USER")
+    DB_PASSWORD: str = os.getenv("DB_PASSWORD")
+    DB_HOST: str = os.getenv("DB_HOST")
     DB_PORT: str = "5432"
```
[the remark this answers is not in the document]
Comment (Vaish24navi) [reply backend/app/utils/settings/config.py:39]: resolved in following commits and changed port

**Review by niyasrad — APPROVED** [review]
Amazing work, admire the hard work! @Vaish24navi  ❤️ 🥂

File: devops/docker-compose.yaml [hunk 4]
```diff
@@ -20,11 +25,29 @@ services:
...
+      ADMIN_USERS: ${POSTGRES_USER}
+      LISTEN_PORT: ${PGBOUNCER_PORT}
+      AUTH_TYPE: scram-sha-256
+      POOL_MODE: transaction
+    ports:
+      - "${PGBOUNCER_PORT}:${PGBOUNCER_PORT}"
+    depends_on:
+      - db
+    networks:
+      - backend
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply devops/docker-compose.yaml:46]: Thank you for the suggestion. We will take this in a future release along with helm/ K8s based functionalities.

File: devops/docker-compose.yaml [hunk 5]
```diff
@@ -4,13 +4,18 @@
       context: ..
       dockerfile: devops/Dockerfile
     environment:
-      DB_HOST: db
+      DB_HOST: pgbouncer
+      DB_NAME: ${POSTGRES_DB}
+      DB_USER: ${POSTGRES_USER}
+      DB_PASSWORD: ${POSTGRES_PASSWORD}
+      DB_PORT: ${PGBOUNCER_PORT}
```
[the remark this answers is not in the document]
Comment (niyasrad) [reply devops/docker-compose.yaml:11]: As mentioned above, we acknowledge the importance of this, and will take this up in future releases. Thank you.

**Pull request merged**